### PR TITLE
Fix pairwise for type-unstable corner case function

### PR DESCRIPTION
**Pull request by nalimilan**
`promote_type` is not a completely correct way of computing an upper bound for the return eltype. Use the same strategy as `map` and `broadcast` in Base instead (https://github.com/JuliaLang/julia/pull/30485), but with `promote_eltype` rather than `promote_typejoin`. We can still use `typejoin_union_tuple` since promotion does not happen inside tuple types.

On Julia versions before 1.6 we would have to copy the full definition of `typejoin_union_tuple`, which is quite complex, so instead fall back to inferring eltype `Any` (inference wasn't so good anyway on these versions).

Fixes https://github.com/JuliaStats/StatsBase.jl/issues/771.

### CODE REVIEW

File: src/pairwise.jl [hunk 1]
```diff
@@ -122,6 +122,32 @@ function _pairwise!(f, dest::AbstractMatrix, x, y;
...
+        elseif T isa Union
+            return promote_type(promote_type_union(T.a), promote_type_union(T.b))
+        elseif T <: Tuple
+            return typejoin_union_tuple(T)
+        else
+            return T
+        end
+    end
+else
+    promote_type_union(::Type) = Any
```
Comment (bkamins) [inline src/pairwise.jl:148]: do we need to be that defensive? If passed type is not a union then I guess we can return it - right?

File: src/pairwise.jl [hunk 2]
```diff
@@ -122,6 +122,32 @@ function _pairwise!(f, dest::AbstractMatrix, x, y;
...
+    if isdefined(Base, :typejoin_union_tuple)
+        using Base: typejoin_union_tuple
+    else
+        using Base.Broadcast: typejoin_union_tuple
+    end
+    # Identical to `Base.promote_typejoin` except that it uses `promote_type`
+    # instead of `typejoin`
+    function promote_type_union(::Type{T}) where T
+        if T === Union{}
+            return Union{}
```
Comment (bkamins) [inline src/pairwise.jl:136]: I would make sure we cover all cases with tests.

File: src/pairwise.jl (hunk 1 above)
Comment (nalimilan) [reply src/pairwise.jl]: Good point. Actually the only case that would be hard to support is `Tuple` types. I've refactored the code to improve this.

File: src/pairwise.jl [hunk 3]
```diff
@@ -122,6 +122,33 @@ function _pairwise!(f, dest::AbstractMatrix, x, y;
...
+    if T === Union{}
+        return Union{}
+    elseif T isa UnionAll
+        return Any # TODO: compute more precise bounds
+    elseif T isa Union
+        return promote_type(promote_type_union(T.a), promote_type_union(T.b))
+    elseif T <: Tuple
+        return typejoin_union_tuple(T)
+    else
+        return T
```
Comment (bkamins) [inline src/pairwise.jl:148]: Is line 148 covered in tests?

Comment (nalimilan) [reply src/pairwise.jl:148]: It's covered as `promote_type_union` calls itself recursively, but I've added two tests for clarity.

**Review by bkamins — APPROVED** [review]

**Pull request merged**